File: cplusplus_bazel/binary_heap_test.cc
#include "binary_heap.h"

#include <gtest/gtest.h>

TEST(TestBinaryHeap, NewHeap) {
  BinaryHeap<int, double> heap;
  EXPECT_EQ(heap.size(), 0);
}

TEST(TestBinaryHeap, AddRaw) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.AddRaw(1, 10).ok());
  EXPECT_TRUE(heap.AddRaw(2, 20).ok());
  EXPECT_TRUE(heap.AddRaw(3, 30).ok());
  EXPECT_TRUE(absl::IsInvalidArgument(heap.AddRaw(3, 40)));
  EXPECT_EQ(heap.size(), 3);
  EXPECT_TRUE(heap.CheckRep().ok());
}

TEST(TestBinaryHeap, IsHeapOrder) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.AddRaw(1, 10).ok());
  EXPECT_TRUE(heap.AddRaw(2, 20).ok());
  EXPECT_TRUE(heap.AddRaw(3, 30).ok());
  EXPECT_TRUE(heap.CheckRep().ok());
  EXPECT_TRUE(heap.IsHeapOrder());
}

TEST(TestBinaryHeap, NotHeapOrder) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.AddRaw(1, 30).ok());
  EXPECT_TRUE(heap.AddRaw(2, 20).ok());
  EXPECT_TRUE(heap.AddRaw(3, 10).ok());
  EXPECT_TRUE(heap.CheckRep().ok());
  EXPECT_FALSE(heap.IsHeapOrder());
}

TEST(TestBinaryHeap, Build) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.AddRaw(1, 30).ok());
  EXPECT_TRUE(heap.AddRaw(2, 20).ok());
  EXPECT_TRUE(heap.AddRaw(3, 10).ok());
  heap.Build();
  EXPECT_TRUE(heap.CheckRep().ok());
  EXPECT_TRUE(heap.IsHeapOrder());
}

TEST(TestBinaryHeap, Add) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.Add(1, 30).ok());
  EXPECT_TRUE(heap.Add(2, 20).ok());
  EXPECT_TRUE(heap.Add(3, 10).ok());
  EXPECT_TRUE(heap.CheckRep().ok());
  EXPECT_TRUE(heap.IsHeapOrder());
}

TEST(TestBinaryHeap, Pop) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.Add(1, 30).ok());
  EXPECT_TRUE(heap.Add(2, 20).ok());
  EXPECT_TRUE(heap.Add(3, 10).ok());

  absl::optional<HeapEntry<int, double>> pop_result = heap.Pop();
  EXPECT_EQ(pop_result.value().object, 3);
  EXPECT_EQ(pop_result.value().value, 10);
  EXPECT_EQ(heap.size(), 2);
  EXPECT_TRUE(heap.CheckRep().ok());

  pop_result = heap.Pop();  
  EXPECT_EQ(pop_result.value().object, 2);
  EXPECT_EQ(pop_result.value().value, 20);
  EXPECT_EQ(heap.size(), 1);
  EXPECT_TRUE(heap.CheckRep().ok());

  pop_result = heap.Pop();  
  EXPECT_EQ(pop_result.value().object, 1);
  EXPECT_EQ(pop_result.value().value, 30);
  EXPECT_TRUE(heap.CheckRep().ok());

  pop_result = heap.Pop();
  EXPECT_FALSE(pop_result.has_value());
  EXPECT_TRUE(heap.CheckRep().ok());
}

TEST(TestBinaryHeap, GetValue) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.Add(1, 30).ok());
  EXPECT_TRUE(heap.Add(2, 20).ok());
  EXPECT_TRUE(heap.Add(3, 10).ok());

  EXPECT_EQ(heap.GetValue(1).value(), 30);
  EXPECT_EQ(heap.GetValue(2).value(), 20);
  EXPECT_EQ(heap.GetValue(3).value(), 10);
  EXPECT_FALSE(heap.GetValue(4).has_value());
}

TEST(TestBinaryHeap, ChangeValue) {
  BinaryHeap<int, double> heap;
  EXPECT_TRUE(heap.Add(1, 30).ok());
  EXPECT_TRUE(heap.Add(2, 20).ok());
  EXPECT_TRUE(heap.Add(3, 10).ok());

  EXPECT_TRUE(heap.ChangeValue(1, 0).ok());
  EXPECT_TRUE(heap.CheckRep().ok());
  EXPECT_TRUE(heap.IsHeapOrder());

  EXPECT_EQ(heap.GetValue(1).value(), 0);
  EXPECT_EQ(heap.GetValue(2).value(), 20);
  EXPECT_EQ(heap.GetValue(3).value(), 10);
}

File: java_bazel/javatests/dijkstra/BinaryHeapTest.java
package dijkstra;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth8.assertThat;

@RunWith(JUnit4.class)
public class BinaryHeapTest {
  @Test
  public void emptyBinaryHeap() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();

    heap.checkRep();
    assertThat(heap.size()).isEqualTo(0);
    assertThat(heap.getValue("A")).isEmpty();
  }

  @Test
  public void addRaw() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.addRaw("A", 10.0);
    heap.addRaw("B", 20.0);

    heap.checkRep();
    assertThat(heap.isHeapOrder()).isTrue();
    assertThat(heap.size()).isEqualTo(2);
    assertThat(heap.getValue("A")).hasValue(10.0);
    assertThat(heap.getValue("B")).hasValue(20.0);
    assertThat(heap.getValue("C")).isEmpty();
  }

  @Test
  public void isHeapOrder_false() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.addRaw("A", 10.0);
    heap.addRaw("B", 20.0);
    heap.addRaw("C", 5.0);

    assertThat(heap.isHeapOrder()).isFalse();
  }

  @Test
  public void swap() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.addRaw("A", 10.0);
    heap.addRaw("B", 20.0);
    heap.addRaw("C", 5.0);

    heap.checkRep();
    heap.swap(1,2);
    heap.checkRep();
  }

  @Test
  public void build() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.addRaw("A", 10.0);
    heap.addRaw("B", 20.0);
    heap.addRaw("C", 5.0);

    heap.checkRep();
    assertThat(heap.isHeapOrder()).isFalse();
    heap.build();
    heap.checkRep();
    assertThat(heap.isHeapOrder()).isTrue();
  }

  @Test
  public void add() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.add("A", 10.0);
    heap.add("B", 20.0);
    heap.add("C", 5.0);

    assertThat(heap.isHeapOrder()).isTrue();
  }

  @Test
  public void deleteMin() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.add("A", 10.0);
    heap.add("B", 20.0);
    heap.add("C", 5.0);

    assertThat(heap.deleteMin()).hasValue("C");
    assertThat(heap.deleteMin()).hasValue("A");
    assertThat(heap.deleteMin()).hasValue("B");
    assertThat(heap.deleteMin()).isEmpty();
  }

  @Test
  public void changeValue() {
    BinaryHeap<String, Double> heap = new BinaryHeap<>();
    heap.add("A", 10.0);
    heap.add("B", 20.0);
    heap.add("C", 5.0);
    heap.changeValue("B", 1.0);

    assertThat(heap.deleteMin()).hasValue("B");
    assertThat(heap.deleteMin()).hasValue("C");
    assertThat(heap.deleteMin()).hasValue("A");
    assertThat(heap.deleteMin()).isEmpty();
  }
}
